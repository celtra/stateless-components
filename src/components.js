import Checkbox from '@/stateless/checkbox.vue'
import DialogButton from '@/stateless/dialog_button.vue'
import Dialog from '@/stateless/dialog.vue'
import DropArea from '@/stateless/drop_area.vue'
import FileUploadRequirements from '@/stateless/file_upload_requirements.vue'
import FileUpload from '@/stateless/file_upload.vue'
import ImageList from '@/stateless/image_list.vue'
import Input from '@/stateless/input.vue'
import Multiselect from '@/stateless/multiselect.vue'
import RadioButton from '@/stateless/radiobutton.vue'
import Selectbox from '@/stateless/selectbox.vue'
import SupportText from '@/stateless/support_text.vue'
import Icon from '@/stateless/icon.vue'
import DefaultList from '@/stateless/DefaultList.vue'
import DefaultListItem from '@/stateless/DefaultListItem.vue'
import ScrollableList from '@/stateless/ScrollableList.vue'
import Typeahead from '@/stateless/Typeahead.vue'
import TypeaheadMultiselect from '@/stateless/TypeaheadMultiselect.vue'
import ColumnFilter from '@/stateless/ColumnFilter.vue'
import Toast from '@/stateless/Toast.vue'
import PieChart from '@/stateless/PieChart.vue'
import InlineDialog from '@/stateless/InlineDialog.vue'
import FilterChip from '@/stateless/FilterChip.vue'
import FilterChipWithMultiselect from '@/stateless/FilterChipWithMultiSelect.vue'

const defaultItems = [
    { id: '1', label: "Something", metadata: 'zan.kusterle@gmail.com', icon: 'plus' },
    { id: '2', label: "Lorem" },
    { id: '3', label: "Ipsum", metadata: 'someone@lorem.ipsum' },
    { id: '4', label: "A", metadata: 'someone@lorem.ipsum' },
    { id: '5', label: "B", metadata: 'someone@lorem.ipsum' },
    { id: '6', label: "C", metadata: 'someone@lorem.ipsum' },
    { id: '7', label: "D", metadata: 'someone@lorem.ipsum' },
    { id: '8', label: "E", metadata: 'someone@lorem.ipsum' },
    { id: '9', label: "F", metadata: 'someone@lorem.ipsum' },
    { id: '10', label: "G", metadata: 'someone@lorem.ipsum' },
]

export default {
    Checkbox: {
        component: Checkbox,
        defaultProps: {
            value: false,
        },
    },
    DialogButton: {
        component: DialogButton,
    },
    Dialog: {
        component: Dialog,
    },
    DropArea: {
        component: DropArea,
    },
    FileUploadRequirements: {
        component: FileUploadRequirements,
        defaultProps: {
            requirements: [
                { name: 'Format', value: 'PNG, JPG' },
                { name: 'Size', value: 'max. 3MB' },
            ],
        },
    },
    FileUpload: {
        component: FileUpload,
        defaultProps: {
            file: {},
            uploadUrl: '',
        },
    },
    ImageList: {
        component: ImageList,
        defaultProps: {
            images: [
                {},
                {},
            ],
        },
    },
    Input: {
        component: Input,
        rootCss: {
            width: '300px',
        },
        defaultProps: {
            label: 'Something',
            theme: 'light',
        },
    },
    Multiselect: {
        component: Multiselect,
        defaultProps: {
            options: defaultItems,
            value: [],
        },
    },
    RadioButton: {
        component: RadioButton,
        defaultProps: {
            value: 'something',
            selectedValue: 'something',
        },
    },
    Selectbox: {
        component: Selectbox,
        rootCss: {
            width: '300px',
        },
        defaultProps: {
            groups: [{ options: defaultItems }],
            isSearchable: true,
            isUnselectable: true,
            theme: 'light',
            size: 'condensed',
            label: 'Something',
        },
    },
    SupportText: {
        component: SupportText,
        defaultProps: {
            text: 'Lorem Ipsum',
        },
    },
    Icon: {
        component: Icon,
    },
    DefaultList: {
        component: DefaultList,
    },
    DefaultListItem: {
        component: DefaultListItem,
    },
    ScrollableList: {
        component: ScrollableList,
    },
    Typeahead: {
        component: Typeahead,
        defaultProps: {
            label: 'Something',
            value: 'Lorem',
            getSuggestions: () => [
                { id: '1', label: "Something", metadata: 'zan.kusterle@gmail.com', icon: 'plus' },
                { id: '2', label: "Lorem" },
                { id: '3', label: "Ipsum", metadata: 'someone@lorem.ipsum' },
            ],
        },
    },
    TypeaheadMultiselect: {
        component: TypeaheadMultiselect,
    },
    ColumnFilter: {
        component: ColumnFilter,
        rootCss: {
            width: '400px',
            boxShadow: '1px 2px 5px 0 rgba(0,0,0,0.25)',
            height: 'fit-content',
        },
        defaultProps: {
            columns: [
                {
                    items: [
                        { id: 'creativeName', label: 'Creative name', type: 'String' },
                        { id: 'unit', label: 'Unit', type: 'String' },
                        { id: 'pageViews', label: 'Page views', type: 'Number' },
                    ],
                },
                {
                    label: 'Campaign level filtering',
                    items: [
                        { id: 'creative', label: 'Creative', type: 'String' },
                        { id: 'placement', label: 'Placement', type: 'String' },
                        {
                            id: 'supplier',
                            label: 'Supplier',
                            type: 'Multiselect',
                            props: {
                                options: [
                                    { id: 'banner', label: 'Banner', metadata: 'Something', ratio: 0.9 },
                                    { id: 'expandableBanner', label: 'ExpandableBanner', metadata: 'Something', ratio: 0.2 },
                                ],
                                getOptions (v) {
                                    return new Promise((resolve, reject) => {
                                        resolve([
                                            { id: 'interstitial', label: 'Interstitial' },
                                        ])
                                    })
                                },
                                autoReorder: false,
                            },
                        },
                    ],
                },
            ],
        },
    },
    Toast: {
        component: Toast,
    },
    PieChart: {
        component: PieChart,
    },
    InlineDialog: {
        component: InlineDialog,
    },
    FilterChip: {
        component: FilterChip,
        defaultProps: {
            click: () => {
                console.log('click on filter chip')
            },
            selectionCounter: {
                selected: 0, all: 4,
            },
            label: 'Label',
        },
    },
    FilterChipWithMultiSelect: {
        component: FilterChipWithMultiselect,
        defaultProps: {
            options: [
                { id: '1', label: '1', metadata: '1' },
                { id: '2', label: '2', metadata: '2' },
                { id: '3', label: '3', metadata: '3' },
                { id: '4', label: '4', metadata: '4' },
                { id: '5', label: '5', metadata: '5' },
                { id: '6', label: '6', metadata: '6' },
                { id: '7', label: '7', metadata: '7' },
                { id: '8', label: '8', metadata: '8' },
                { id: '9', label: '9', metadata: '9' },
            ],
            value: [ '1', '2' ],
            searchableWhenOptionsLengthIsMoreThan: 6,
            label: 'Test Label',
        },
    },
}
