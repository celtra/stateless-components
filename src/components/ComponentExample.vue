<script>
const getModelName = component => component.model && component.model.prop || 'value'
const getModelEvent = component => component.model && component.model.event || 'input'

export default {
    props: {
        component: { type: Object, required: true },
        value: { type: null, required: false },
        setup: { type: Function },
        scopedSlots: { type: Object },
        columnIndex: { type: Number, required: true },
        rowIndex: { type: Number, required: true },
    },
    data () {
        return {
            currentValue: null,
            height: null,
        }
    },
    watch: {
        value (v) {
            this.currentValue = v
        },
    },
    created () {
        const component = this.component
        const modelName = getModelName(component)
        if (typeof this.$attrs[modelName] !== 'undefined') {
            this.currentValue = this.$attrs[modelName]
        } else {
            let defaultValue = component.props && component.props[modelName] && component.props[modelName].default
            if (typeof defaultValue === 'undefined') {
                const type = component.props && component.props[modelName] && component.props[modelName].type
                if (type === Array) {
                    defaultValue = []
                } else if (type === Object) {
                    defaultValue = {}
                } else if (type === String) {
                    defaultValue = ''
                } else {
                    defaultValue = null
                }
            }
            this.currentValue = defaultValue
        }
    },
    mounted () {
        if (this.setup) {
            this.setup(this.$children[0])
        }

        if (this.columnIndex === 1) {
            const styles = window.getComputedStyle(this.$el)
            const margin = parseFloat(styles['marginTop']) + parseFloat(styles['marginBottom'])
            this.$nextTick(() => {
                setTimeout(() => {
                    this.$parent.$emit('height', { rowIndex: this.rowIndex, height: this.$el.clientHeight + margin })
                }, 5)
            })
        }
    },
    methods: {
        updateValue (value) {
            this.currentValue = value
            this.$emit('input', value)
        },
    },
    render (h) {
        const props = {
            ...this.$attrs,
            [getModelName(this.component)]: this.currentValue,
        }

        let slot = this.slot ? this.slot.bind(props)(h) : null
        if (typeof slot === 'string') {
            slot = this._v(slot)
        }

        const scopedSlots = {}
        if (this.scopedSlots) {
            for (const name in this.scopedSlots) {
                scopedSlots[name] = () => this.scopedSlots[name](h)
            }
        }

        return h(this.component, {
            props: props,
            scopedSlots: scopedSlots,
            class: this.$style.component,
            on: {
                [getModelEvent(this.component)]: (value) => {
                    this.updateValue(value)
                },
            },
        }, slot ? [slot] : [])
    },
}
</script>

<style lang="less" module>
.component {
    width: 100% !important;
    margin: 0 !important;
    height: auto !important;
}
</style>
