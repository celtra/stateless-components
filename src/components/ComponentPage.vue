<template>
    <div :class="`main--${theme}`" class="main">
        <div class="header">
            <checkbox :is-toggle="true" v-model="boundsVisible" :theme="theme" style="margin-top: 0; height: auto;">Bounds</checkbox>
            <checkbox :is-toggle="true" v-model="isEventsListOpen" :theme="theme" style="margin-left: 15px; margin-top: 0; height: auto;">Events</checkbox>
            <checkbox :is-toggle="true" :disabled="component.forceValueSync" v-model="syncValue" :theme="theme" style="margin-left: 15px; margin-top: 0; height: auto;">Sync model</checkbox>
            <div class="props-info">
                <chip
                    v-for="(values, name) in componentVariations"
                    v-if="name !== 'value'"
                    :is-removable="name in filters"
                    :key="name" :label="kebabCase(name).toUpperCase()" :metadata="name in filters ? filters[name] + '' : null"
                    :is-active="name in filters"
                    :theme="theme"
                    class="prop-info"
                    @click="cycleFilter(name)"
                    @remove="removeFilter(name)"
                />
            </div>
        </div>

        <div :style="isEventsListOpen ? { paddingLeft: '370px' } : {}" class="component-view">
            <component-variations :use-sync-value="syncValue || component.forceValueSync || false" :component="component" :filters="filters" :show-bounding-boxes="boundsVisible" />
        </div>

        <div class="sidebar browse">
            <default-list :items="componentNames.map(name => ({ id: name, label: name }))" :theme="theme" @select="$router.push({ name: 'ComponentPage', params: { component: $event.label } })">
                <div slot-scope="{ item }" :class="{ 'sidebar-item--active': item.id === name }" class="sidebar-item">
                    {{ item.label }}
                </div>
            </default-list>
        </div>

        <div v-if="isEventsListOpen" class="sidebar events">
            <default-list :items="events.slice().reverse().map((event, index) => ({ id: index, event: event }))" :theme="theme">
                <div slot-scope="{ item }" class="sidebar-item" @click="currentEventIndex = item.id">
                    <p v-if="item.event" class="event-name">{{ item.event.componentName }}/{{ item.event.name }}</p>
                    <div v-if="currentEventIndex === item.id" class="event-payload">
                        <template v-if="item.event.payload.length > 0">
                            <template v-if="typeof item.event.payload[0] === 'object'">
                                <p v-for="(value, key) in item.event.payload[0]" :key="key"><b>{{ key }}:</b> {{ value }}</p>
                            </template>
                            <template v-else>
                                {{ item.event.payload[0] }}
                            </template>
                        </template>
                        <template v-else>
                            <p><b>no data</b></p>
                        </template>
                    </div>
                </div>
            </default-list>
        </div>

    </div>
</template>

<script>
import Vue from 'vue'
import { kebabCase } from 'lodash'
import '@/stateless/define_helpers'
import * as library from '../library.js'
import Chip from '@/stateless/Chip.vue'
import Checkbox from '@/stateless/checkbox.vue'
import Icon from '@/stateless/icon.vue'
import DefaultList from '@/stateless/DefaultList.vue'
import ComponentVariations from './ComponentVariations.vue'

export default {
    components: {
        Chip,
        Checkbox,
        Icon,
        DefaultList,
        ComponentVariations,
    },
    data () {
        return {
            name: null,
            events: [],
            currentEventIndex: null,
            showThemeToggle: true,
            isThemeLight: true,
            boundsVisibleData: false,
            filters: {},
            isEventsListOpenData: false,
            syncValueData: false,
        }
    },
    computed: {
        syncValue: {
            get () {
                return this.component && this.component.forceValueSync || this.syncValueData
            },
            set (v) {
                localStorage.setItem('syncValue', v.toString())
                this.syncValueData = v
            },
        },
        boundsVisible: {
            get () {
                return this.boundsVisibleData
            },
            set (v) {
                localStorage.setItem('boundsVisible', v.toString())
                this.boundsVisibleData = v
            },
        },
        isEventsListOpen: {
            get () {
                return this.isEventsListOpenData
            },
            set (v) {
                localStorage.setItem('isEventsListOpen', v.toString())
                this.isEventsListOpenData = v
            },
        },
        componentNames () {
            return Object.values(library).filter(x => x.usecases && !x.hasAbsolutePosition).map(x => x.metaName).sort()
        },
        component () {
            if (this.name === null) {
                return null
            }
            return library[this.name]
        },
        componentVariations () {
            const variations = this.component && { ...this.component.variations } || {}
            if (this.component.usecases[0].name) {
                variations.name = this.component.usecases.map(usecase => usecase.name)
            }
            return variations
        },
        theme () {
            return this.filters.theme === 'dark' ? 'dark' : 'light'
        },
    },
    watch: {
        '$route.params.component' (name) {
            this.setupComponent()
        },
        '$route.params.filters' (value) {
            this.setupFilters()
        },
    },
    created () {
        this.syncValue = localStorage.getItem('syncValue') === 'true' ? true : false
        this.isEventsListOpen = localStorage.getItem('isEventsListOpen') === 'true' ? true : false
        this.boundsVisible = localStorage.getItem('boundsVisible') === 'true' ? true : false

        this.kebabCase = kebabCase
        const original = Vue.prototype.$emit
        const logEvent = this.logEvent
        Vue.prototype.$emit = function (...args) {
            const componentName = this.$options.name || this.$options._componentTag || 'Root'
            const res = original.apply(this, args)
            logEvent(kebabCase(componentName), args[0], args.slice(1))
            return res
        }
        this.setupComponent()
        this.setupFilters()
    },
    methods: {
        setupComponent () {
            const name = this.$route.params.component
            const component = library[name]
            if (!component) {
                throw `Component ${name} does not exist!`
            }
            this.name = name
        },
        setupFilters () {
            if (this.$route.params.filters) {
                const newFilters = {}
                const filterValues = this.$route.params.filters.split('&')
                for (const filterValue of filterValues) {
                    const parts = filterValue.split('=')
                    if (parts.length === 2) {
                        const name = parts[0], value = parts[1]
                        const propData = this.component.props[name]
                        if (propData) {
                            let newValue = value
                            if (propData.type === Boolean) {
                                newValue = value === 'true'
                            }
                            newFilters[name] = newValue
                        }
                    }
                }
                this.filters = newFilters
            }
        },
        logEvent (componentName, eventName, payload) {
            if (this.currentEventIndex !== null) {
                this.currentEventIndex ++
            }
            this.events.push({ componentName: componentName, name: eventName, payload: payload })
        },
        cycleFilter (name) {
            const currentValue = this.filters[name]
            const values = this.componentVariations[name]
            const currentIndex = values.indexOf(currentValue)
            const newIndex = (currentIndex + 1) % values.length
            Vue.set(this.filters, name, values[newIndex])
            this.updateUrlFilters()
        },
        removeFilter (name) {
            Vue.delete(this.filters, name)
            delete this.filters[name]
            this.updateUrlFilters()
        },
        updateUrlFilters () {
            const value = Object.keys(this.filters).sort().map(name => `${name}=${this.filters[name]}`).join('&')
            this.$router.replace({ name: 'ComponentPage', params: { component: this.name, filters: value || null } })
        },
    },
}
</script>

<style lang="less" scoped>
@dark-theme: #1f1f2c;

.main {
    height: 100%;

    &--light {
        background-color: white;
        color: black;

        .header {
            background-color: white;
        }
    }

    &--dark {
        background-color: @dark-theme;

        .header {
            color: black;
            background-color: @dark-theme;
        }

        .sidebar-item {
            color: white;
        }
    }
}

.header {
    display: flex;
    align-items: center;
    padding-left: 20px;
    box-sizing: border-box;
    width: calc(~'100% - 20px');

    user-select: none;
    height: 50px;
    position: fixed;
    top: 0;
    left: 0;
    border-bottom: 1px solid rgba(122, 122, 122, 0.1);

    .theme-toggle {
        margin-top: 0;
        margin-left: 50px;
    }

    .header-icon {
        cursor: pointer;
    }

    .view-toggle {
        font-size: 14px;
        margin-left: 10px;
        cursor: pointer;
    }

    .header-left {
        display: flex;
        align-items: center;
    }
}

.component-view {
    padding-left: 155px;
    padding-top: 70px;
    overflow-y: auto;
    width: 100%;
    height: 100%;
    display: flex;
    flex-flow: column;
    box-sizing: border-box;
    transition: padding-left 500ms ease-out;
    z-index: 100;
    user-select: none;
}

.props-info {
    display: flex;
    overflow-x: hidden;
    margin-left: 10px;
}

.prop-info {
    margin: 5px;
}

.sidebar {
    position: fixed;
    top: 70px;
    height: 100%;
    display: flex;
    flex-direction: column;
    z-index: 10;
}

.browse {
    width: 150px;
}

.events {
    left: 150px;
    top: 50px;
    width: 200px;
    border: 1px solid rgba(0, 0, 0, 0.1);
    border-width: 0 1px;
    animation: fadeIn 500ms linear;
    animation-fill-mode: forwards;
    opacity: 0;
    animation-delay: 300ms;
}

@keyframes fadeIn {
    from { opacity: 0; }
    to { opacity: 1; }
}

.sidebar-item {
    padding: 10px 0px;
    font-size: 15px;
    cursor: pointer;
    display: flex;
    flex-direction: column;
    align-items: center;

    &--active {
        font-weight: bold;
    }
}

.event-payload {
    margin-top: 5px;
    padding-left: 20px;
    padding-top: 5px;

    > p {
        margin: 0;
        font-size: 15px;

        > b {
            text-transform: uppercase;
            font-size: 12px;
        }
    }
}
</style>
