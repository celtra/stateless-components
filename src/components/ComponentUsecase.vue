<script>
const getModelName = component => component.model && component.model.prop || 'value'
const getModelEvent = component => component.model && component.model.event || 'input'

export default {
    props: {
        component: { type: Object, required: true },
        usecase: { type: Object, required: true },
    },
    data () {
        return {
            value: null,
        }
    },
    created () {
        const component = this.component
        const modelName = getModelName(component)
        if (typeof this.usecase.data.value !== 'undefined') {
            this.value = this.usecase.data.value
        } else {
            let defaultValue = component.props && component.props[modelName] && component.props[modelName].default
            if (typeof defaultValue === 'undefined') {
                const type = component.props && component.props[modelName] && component.props[modelName].type
                if (type === Array) {
                    defaultValue = []
                } else if (type === Object) {
                    defaultValue = {}
                } else if (type === String) {
                    defaultValue = ''
                } else {
                    defaultValue = null
                }
            }
            this.value = defaultValue
        }
    },
    methods: {
        updateValue (value) {
            this.value = value
        },
    },
    render (h) {
        const props = {
            ...this.usecase.data,
            [getModelName(this.component)]: this.value,
        }

        let slot = this.usecase.data.slot ? this.usecase.data.slot.bind(props)(h) : null
        if (typeof slot === 'string') {
            slot = this._v(slot)
        }

        return h(this.component, {
            props: props,
            on: {
                [getModelEvent(this.component)]: (value) => {
                    this.updateValue(value)
                },
            },
        }, slot ? [slot] : [])
    },
}
</script>
