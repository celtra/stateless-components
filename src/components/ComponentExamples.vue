<template>
    <div>
        <div :class="$style.filters">
            <icon :class="$style.resetFilters" name="x-bold" @click="$emit('reset-filters')" />
            <chip
                v-for="(values, name) in valuesByName"
                v-if="name !== 'value'"
                :is-removable="name in filters"
                :key="name"
                :label="getFilterTitle(name)"
                :is-active="name in filters"
                :theme="filters.theme || 'light'"
                :class="$style.propInfo"
                @click="cycleFilter(name)"
                @remove="removeFilter(name)"
            />
        </div>

        <div v-for="(columns, rowIndex) in rows" :key="rowIndex">
            <examples-table :columns="columns">
                <div slot-scope="item" :class="$style.boundingBox" class="bounding-box">
                    <component-example
                        :class="$style.component"
                        :key="item.name"
                        :component="component"
                        v-bind="item"
                        :theme="filters.theme || item.theme || 'light'"
                        :value="syncValue"
                        @input="(v) => useSyncValue && (syncValue = v)">
                    </component-example>
                </div>
            </examples-table>
        </div>
    </div>
</template>

<script>
import { getFlatUsecases, getFlatVariations } from '../component_utils'
import { kebabCase } from 'lodash'
import ExamplesTable from './ExamplesTable.vue'
import ComponentExample from './ComponentExample.vue'
import { maxBy, pickBy } from 'lodash'
import { getPropTitle } from './utils'
import { filter } from '../stateless/items_utils'
import Icon from '@/stateless/icon.vue'
import Chip from '@/stateless/Chip.vue'

export default {
    components: {
        ExamplesTable,
        ComponentExample,
        Icon,
        Chip,
    },
    props: {
        component: { type: Object, required: true },
        filters: { type: Object, default: () => ({}) },
        showBoundingBoxes: { type: Boolean, default: false },
        useSyncValue: { type: Boolean, default: false },
    },
    data () {
        return {
            syncValue: null,
        }
    },
    computed: {
        modelName () {
            return this.component.model && this.component.model.prop || 'value'
        },
        valuesByName () {
            // copied from ComponentPage.componentVariations
            const variations = this.component && { ...this.component.variations } || {}
            if (this.component.usecases[0].name) {
                variations.usecaseName = this.component.usecases.map(usecase => usecase.name)
            }

            return variations
        },
        splitByProp () {
            const props = this.valuesByName ? Object.keys(this.valuesByName).filter(key => !(key in this.filters)) : []

            const relevances = {
                theme: 2,
                size: 1,
            }
            const names = props.filter(x => !Object.keys(this.filters).concat([this.modelName]).includes(x))
            let columnProp = maxBy(names.filter(name => this.valuesByName[name].length <= 3 && name !== 'usecaseName'), x => {
                const relevance = (relevances[x] || 0)
                const length = this.valuesByName[x].length
                return relevance * 1000 + length
            })
            let rowProp = maxBy(names.filter(name => name !== columnProp && name !== 'usecaseName'), x => {
                const relevance = (relevances[x] || 0)
                const length = this.valuesByName[x].length
                return relevance * 1000 + length
            })

            let product = 1
            Object.values(this.valuesByName).forEach(values => product *= values.length)

            if (columnProp) {
                if (Math.round(product / this.valuesByName[columnProp].length) < 4) {
                    columnProp = null
                } else {
                    product /= this.valuesByName[columnProp].length
                }
            }

            if (rowProp) {
                if (Math.round(product / this.valuesByName[rowProp].length) < 4) {
                    rowProp = null
                }
            }

            const remainingValues = _.pickBy(this.valuesByName, (value, name) => {
                return !Object.keys(this.filters).concat([this.modelName, columnProp, rowProp]).includes(name)
            })

            const flat = []
            for (const variation of getFlatVariations(remainingValues)) {
                const usecase = this.component.usecases.find(x => x.name === (variation.usecaseName || this.filters.usecaseName))
                const variationSuffix = Object.keys(variation).map(key => {
                    const propData = this.component.props[key]
                    if (!propData) {
                        return null
                    }
                    if (propData.type === Boolean) {
                        return variation[key] ? key : null
                    }
                }).filter(x => x).join(', ')
                flat.push({
                    name: this.filters.usecaseName ? null :`${usecase.name} ${variationSuffix}`,
                    variation: { ...variation, ...usecase },
                })
            }

            return {
                column: columnProp,
                row: rowProp,
                flat: flat,
            }
        },
        rows () {
            const themesCss = {
                white: { backgroundColor: 'white', color: 'black' },
                light: { backgroundColor: '#f2f2f3', color: 'black' },
                dark: { backgroundColor: '#1f1f2c', color: 'white' },
            }

            const rowValues = this.splitByProp.row ? this.valuesByName[this.splitByProp.row] : [null]
            const columnValues = this.splitByProp.column ? this.valuesByName[this.splitByProp.column] : [null]
            return rowValues.map((rowValue, rowIndex) => {
                const themeCss = this.filters.theme && themesCss[this.filters.theme] || this.splitByProp.row === 'theme' && themesCss[rowValue]
                const columnItems = [
                    {
                        title: getPropTitle(this.splitByProp.row, rowValue, true),
                        content: this.splitByProp.flat.map(usecase => usecase.name),
                        themeCss: themeCss || themesCss.light,
                    },
                    ...columnValues.map((columnValue, index) => {
                        return {
                            title: getPropTitle(this.splitByProp.column, columnValue),
                            content: this.splitByProp.flat.map(x => ({
                                ...x.variation,
                                ...this.filters,
                                [this.splitByProp.row]: rowValue,
                                [this.splitByProp.column]: columnValue,
                            })),
                            themeCss: themeCss || this.splitByProp.column === 'theme' && themesCss[columnValue] || themesCss.light,
                        }
                    }),
                ]
                return columnItems
            })
        },
    },
    methods: {
        cycleFilter (name) {
            const currentValue = this.filters[name]
            const values = this.valuesByName[name]
            const currentIndex = values.indexOf(currentValue)
            const newIndex = (currentIndex + 1) % values.length
            this.$emit('set-filter', name, values[newIndex])
        },
        removeFilter (name) {
            this.$emit('unset-filter', name)
        },
        getFilterTitle (name) {
            if (name === 'usecaseName') {
                return name in this.filters ? this.filters[name].toUpperCase() : 'NAME'
            }
            return getPropTitle(name, this.filters[name], true)
        },
    },
}
</script>

<style lang="less" module>
.component {

}

.boundingBox {
}

.filters {
    display: flex;
    align-items: center;
    margin-bottom: 25px;
    background-color: rgba(122, 122, 122, 0.2);
    padding: 10px;
    border-radius: 5px;
}

.resetFilters {
    color: rgba(122, 122, 122, 0.7);
    cursor: pointer;
}

.propInfo {
    margin-left: 10px;
}
</style>
