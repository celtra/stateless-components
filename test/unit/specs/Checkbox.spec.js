import Checkbox from '@/stateless/checkbox.vue'
import Vue from 'vue'
import components from '@/components'

const Constructor = Vue.extend(Checkbox)
const vm = new Constructor({
    propsData: components.Checkbox.defaultProps,
}).$mount()

describe('Checkbox', () => {
    it('should match the snapshot', () => {
        expect(vm.$el).toMatchSnapshot()
    })
})
