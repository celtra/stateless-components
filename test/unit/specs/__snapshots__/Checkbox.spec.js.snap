// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox computed usecase-0__disabled-false_value-false actualTitleText 1`] = `""`;

exports[`Checkbox computed usecase-0__disabled-false_value-false infoText 1`] = `""`;

exports[`Checkbox computed usecase-0__disabled-false_value-false states 1`] = `
Object {
  "checked": false,
  "disabled": false,
  "error": false,
  "focused": false,
  "some": false,
  "warning": false,
}
`;

exports[`Checkbox computed usecase-0__disabled-false_value-null actualTitleText 1`] = `""`;

exports[`Checkbox computed usecase-0__disabled-false_value-null infoText 1`] = `""`;

exports[`Checkbox computed usecase-0__disabled-false_value-null states 1`] = `
Object {
  "checked": false,
  "disabled": false,
  "error": false,
  "focused": false,
  "some": true,
  "warning": false,
}
`;

exports[`Checkbox computed usecase-0__disabled-false_value-true actualTitleText 1`] = `""`;

exports[`Checkbox computed usecase-0__disabled-false_value-true infoText 1`] = `""`;

exports[`Checkbox computed usecase-0__disabled-false_value-true states 1`] = `
Object {
  "checked": true,
  "disabled": false,
  "error": false,
  "focused": false,
  "some": false,
  "warning": false,
}
`;

exports[`Checkbox computed usecase-0__disabled-true_value-false actualTitleText 1`] = `""`;

exports[`Checkbox computed usecase-0__disabled-true_value-false infoText 1`] = `""`;

exports[`Checkbox computed usecase-0__disabled-true_value-false states 1`] = `
Object {
  "checked": false,
  "disabled": true,
  "error": false,
  "focused": false,
  "some": false,
  "warning": false,
}
`;

exports[`Checkbox computed usecase-0__disabled-true_value-null actualTitleText 1`] = `""`;

exports[`Checkbox computed usecase-0__disabled-true_value-null infoText 1`] = `""`;

exports[`Checkbox computed usecase-0__disabled-true_value-null states 1`] = `
Object {
  "checked": false,
  "disabled": true,
  "error": false,
  "focused": false,
  "some": true,
  "warning": false,
}
`;

exports[`Checkbox computed usecase-0__disabled-true_value-true actualTitleText 1`] = `""`;

exports[`Checkbox computed usecase-0__disabled-true_value-true infoText 1`] = `""`;

exports[`Checkbox computed usecase-0__disabled-true_value-true states 1`] = `
Object {
  "checked": true,
  "disabled": true,
  "error": false,
  "focused": false,
  "some": false,
  "warning": false,
}
`;

exports[`Checkbox computed usecase-1__disabled-false_value-false actualTitleText 1`] = `""`;

exports[`Checkbox computed usecase-1__disabled-false_value-false infoText 1`] = `""`;

exports[`Checkbox computed usecase-1__disabled-false_value-false states 1`] = `
Object {
  "checked": false,
  "disabled": false,
  "error": false,
  "focused": false,
  "some": false,
  "warning": false,
}
`;

exports[`Checkbox computed usecase-1__disabled-false_value-null actualTitleText 1`] = `""`;

exports[`Checkbox computed usecase-1__disabled-false_value-null infoText 1`] = `""`;

exports[`Checkbox computed usecase-1__disabled-false_value-null states 1`] = `
Object {
  "checked": false,
  "disabled": false,
  "error": false,
  "focused": false,
  "some": true,
  "warning": false,
}
`;

exports[`Checkbox computed usecase-1__disabled-false_value-true actualTitleText 1`] = `""`;

exports[`Checkbox computed usecase-1__disabled-false_value-true infoText 1`] = `""`;

exports[`Checkbox computed usecase-1__disabled-false_value-true states 1`] = `
Object {
  "checked": true,
  "disabled": false,
  "error": false,
  "focused": false,
  "some": false,
  "warning": false,
}
`;

exports[`Checkbox computed usecase-1__disabled-true_value-false actualTitleText 1`] = `""`;

exports[`Checkbox computed usecase-1__disabled-true_value-false infoText 1`] = `""`;

exports[`Checkbox computed usecase-1__disabled-true_value-false states 1`] = `
Object {
  "checked": false,
  "disabled": true,
  "error": false,
  "focused": false,
  "some": false,
  "warning": false,
}
`;

exports[`Checkbox computed usecase-1__disabled-true_value-null actualTitleText 1`] = `""`;

exports[`Checkbox computed usecase-1__disabled-true_value-null infoText 1`] = `""`;

exports[`Checkbox computed usecase-1__disabled-true_value-null states 1`] = `
Object {
  "checked": false,
  "disabled": true,
  "error": false,
  "focused": false,
  "some": true,
  "warning": false,
}
`;

exports[`Checkbox computed usecase-1__disabled-true_value-true actualTitleText 1`] = `""`;

exports[`Checkbox computed usecase-1__disabled-true_value-true infoText 1`] = `""`;

exports[`Checkbox computed usecase-1__disabled-true_value-true states 1`] = `
Object {
  "checked": true,
  "disabled": true,
  "error": false,
  "focused": false,
  "some": false,
  "warning": false,
}
`;
